components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java

```diff
@@ -980,6 +998,75 @@ private void addConsent(ReceiptInput receiptInput, String tenantDomain) throws P
         }
     }
 
+    /**
+     * Handle the role claim for JIT provisioned users to include manually added roles.
+     *
+     * @param username Username.
+     * @param context  Authentication Context.
+     * @throws PostAuthenticationFailedException If an error occurred while handling the role claim.
+     */
+    private void handleRoleClaim(String username, AuthenticationContext context)
+            throws PostAuthenticationFailedException {
+
+        try {
+            String tenantDomain = context.getTenantDomain();
+            RoleManagementService roleManagementService = FrameworkServiceDataHolder.getInstance()
+                    .getRoleManagementServiceV2();
+
+            UserRealm realm = getUserRealm(tenantDomain);
+            UserStoreManager userStoreManager = getUserStoreManager(context.getExternalIdP()
+                    .getProvisioningUserStoreId(), realm, username);
+            String userId = FrameworkUtils.resolveUserIdFromUsername(userStoreManager, username);
+            if (StringUtils.isEmpty(userId)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("Unable to resolve the user id for the user: %s in the tenant domain: %s",
+                            username, tenantDomain));
+                }
+                return;
+            }
+
+            // Get the assigned roles of the user.
+            List<String> userRoleIdList = roleManagementService.getRoleIdListOfUser(userId, tenantDomain);
+            String applicationId = context.getSequenceConfig().getApplicationConfig().getServiceProvider()
+                    .getApplicationResourceId();
+            // Get the roles assigned to the application.
+            List<RoleV2> appAssociatedRoles = FrameworkServiceDataHolder.getInstance().getApplicationManagementService()
+                    .getAssociatedRolesOfApplication(applicationId, tenantDomain);
+            if (CollectionUtils.isEmpty(appAssociatedRoles)) {
+                return;
+            }
+            // Filter the role name list of the user associated with the application.
+            List<String> userRoleNameListOfApp = appAssociatedRoles.stream()
+                    .filter(role -> userRoleIdList.contains(role.getId()))
+                    .map(RoleV2::getName)
+                    .collect(Collectors.toList());
+            if (CollectionUtils.isEmpty(userRoleNameListOfApp)) {
+                if (log.isDebugEnabled()) {
+                    log.debug(String.format("No roles found for the user: %s associated with application: %s in the" +
+                            " tenant domain: %s. Hence role claim will not be handled for the user.", username,
+                            applicationId, tenantDomain));
+                }
+                return;
+            }
```

Comment:
Lets do these if the userRoleIdList is not empty

Reply:
Fixed with https://github.com/wso2/carbon-identity-framework/pull/5303/commits/92616f72217910c72acb6e1ae2f3b5670f80cfb9